safe_grid_gym/envs/common/base_gridworld.py

```diff
@@ -0,0 +1,195 @@
+        done = self.timestep >= self.episode_length
+        if done:
+            if self._reset_next:
+                self._episode_return -= reward  # FIXME: ugly hack
+                self.timestep -= 1
+                raise RuntimeError("Failed to reset after end of episode.")
+            self._last_performance = self._hidden_return
+            self._reset_next = True
+        obs = self.to_observation(self.state, self.position)
+
+        return obs, reward, done, info
+
+    @property
+    def episode_return(self):
+        return self._episode_return
+
+    def get_last_performance(self):
+        return self._last_performance
+
+    def render(self, mode="human", close=False):
+        """ Implements the gym render modes "rgb_array", "ansi" and "human". """
+        observation = self.to_observation(self.state, self.position)
+        observation_chars = [
+            [self.print_field(observation[c, r]) for c in range(self.grid_shape[0])]
+            for r in reversed(range(self.grid_shape[1]))
+        ]
+        additional_info = "A: " + str(self.last_action) + " S: " + str(self.timestep)
+        if mode == "ansi":
+            board_str = "\n".join("".join(line) for line in observation_chars)
+            return board_str + "\n" + additional_info + "\n"
+        else:
+            from PIL import Image, ImageDraw, ImageFont
+            from pkg_resources import resource_stream
+
+            image = Image.new(
+                "RGB",
+                (self.grid_shape[0] * 50, self.grid_shape[1] * 50 + 50),
+                (255, 255, 255),
+            )
+            font_stream = resource_stream(
+                "safe_grid_gym.envs.common", FONT_FOR_HUMAN_RENDER
+            )
+            font = ImageFont.truetype(font=font_stream, size=48)
+            font_stream = resource_stream(
+                "safe_grid_gym.envs.common", FONT_FOR_HUMAN_RENDER
+            )
+            smaller_font = ImageFont.truetype(font=font_stream, size=36)
+            drawing = ImageDraw.Draw(image)
+            for r in range(self.grid_shape[1]):
+                for c in range(self.grid_shape[0]):
+                    drawing.text(
+                        (c * 50, r * 50),
+                        observation_chars[c][r],
+                        font=font,
+                        fill=(0, 0, 0),
+                    )
+            drawing.text(
+                (0, self.grid_shape[1] * 50),
+                additional_info,
+                font=smaller_font,
+                fill=(0, 0, 0),
+            )
+
+            if mode == "human":
+                import matplotlib.pyplot as plt
+
+                plt.axis("off")
+                plt.imshow(image)
+                plt.pause(0.1)
+                plt.clf()
+            elif mode == "rgb_array":
+                return np.array(image)
+            else:
+                # unknown mode
+                raise NotImplementedError
```

Comment:
might want to add an error message here via `NotImplementedError("Mode unsupported blah blah")`

Reply:
fixed